### PR TITLE
Perform precise check for types warnings in cluster restart tests.

### PR DESCRIPTION
**Pull request by jtibshirani**
Instead of using `WarningsHandler.PERMISSIVE`, we only match warnings
that are due to types removal.

This PR also renames `allowTypeRemovalWarnings` to `allowTypesRemovalWarnings`.

Relates to #37920.

### CODE REVIEW

**Comment by jtibshirani** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/2

**Comment by jtibshirani** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jtibshirani** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestCreateIndexAction.java
Comment (markharwood) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestCreateIndexAction.java]: In a separate PR I was reconsidering this warning message. The deprecation part is clear enough but the solution only mentions setting a flag "to be compatible with the next major version".  Maybe we should be more explicit stating that the user needs to both:
1) Remove the type from the usual `mapping -> type -> properties` hierarchy and 
2) Tell us they've done this with the `include_type_name=false` param 

It's unfortunate that doing step 1 requires users to also do 2 but that's where we are.

**Review by markharwood — COMMENTED** [review]
I left a comment re warning message and build looks to have failed but otherwise +1 for making the warnings checks less permissive.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestCreateIndexAction.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestCreateIndexAction.java]: That does sound clearer to me! Would you be able to make that update on 6.x? Then I could hold this PR until that merges, and update the message here before merging.

**Comment by jtibshirani** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by jtibshirani** [comment]
@markharwood with the new deprecation message, does this now look good to you?

**Comment by jtibshirani** [comment]
After catching up offline, we decided to change the check to verify the message starts with `[types removal]`, as opposed to matching the exact message from 6.7.

**Comment by markharwood** [comment]
I just created https://github.com/elastic/elasticsearch/pull/38310 which adds a new "EsRestTestCase.expectTypesRemovalWarnings" method. As we discussed this just checks for any warning messages starting with `[types removal]` rather than having master rely on a copy of the exact phrasing of messages used in 6.x branch (as they could easily deviate).

**Comment by jtibshirani** [comment]
@markharwood I made an update to use the method `alloTypesRemovalWarnings` from your PR. This should be ready for another look.

**Review by markharwood — APPROVED** [review]
LGTM

**Pull request merged**
